ubiquity/rpc-handler
# PR: include all RPCs by default & allow filter by tracking info

Resolves #22

I have done the following besides the required RPC tracking config options.
- I have added vscode debug configuration files. As we are including tests with all repositories now, this config helps with quick single-click debugging using breakpoints in vscode. This replaces the need for all those `console.logs` that we use to debug. Try this if you haven't. It will make working with code extremely pleasant. I will also start including these debug configs to other repositories too if I don't get an objection here.
- The repository has some heavy code duplications. I have removed all test mocks. Test appear to pass without them.
- Now, tests import from subject source code, not build code as it should be.

I also wanted to add wss/https filter as config option with tracking, but chainlist RPCs have small number of unreliable wss links, so I aborted it. Otherwise, users would have blamed the package for not providing a good wss. But maybe something good in the future to consider.

## QA
Other than passing tests, let's try a modified `tests/script-test.tsx` to see if we get rpcs with correct tracking.   
Using chain: 100 and its Rpcs from chainlist.

1. tracking unspecified: the default behavior of existing users of the package. This should provide all working RPCs ignoring their tracking. 
Script output.
```
config tracking:  undefined
Used Rpcs:
URL: https://rpc.gnosischain.com Tracking:undefined
URL: https://gnosis-pokt.nodies.app Tracking:none
URL: https://rpc.gnosis.gateway.fm Tracking:yes
URL: https://gnosis-mainnet.public.blastapi.io Tracking:limited
URL: https://rpc.ankr.com/gnosis Tracking:limite
```

2. Tracking config: yes
It should use all Rpcs just like the unspecified config for tracking.
Script output:
```
config tracking:  yes
Used Rpcs:
URL: https://rpc.gnosischain.com Tracking:undefined
URL: https://gnosis-pokt.nodies.app Tracking:none
URL: https://rpc.gnosis.gateway.fm Tracking:yes
URL: https://gnosis-mainnet.public.blastapi.io Tracking:limited
URL: https://rpc.ankr.com/gnosis Tracking:limited
URL: https://gnosis.blockpi.network/v1/rpc/public Tracking:limited
```

3. Tracking config: limited
It should use Rpcs with none or limited tracking.
Script output:
```
config tracking:  limited
Used Rpcs:
URL: https://gnosis-pokt.nodies.app Tracking:none
URL: https://gnosis-mainnet.public.blastapi.io Tracking:limited
URL: https://rpc.ankr.com/gnosis Tracking:limited
URL: https://gnosis.blockpi.network/v1/rpc/public Tracking:limited
URL: https://gnosis.drpc.org Tracking:none
URL: https://endpoints.omniatech.io/v1/gnosis/mainnet/public Tracking:none
```

4. Tracking config: none
It should use Rpcs with tracking as "none".
Script output:
```
4. config tracking:  none
Used Rpcs:
URL: https://gnosis-pokt.nodies.app Tracking:none
URL: https://gnosis.drpc.org Tracking:none
URL: https://endpoints.omniatech.io/v1/gnosis/mainnet/public Tracking:none
URL: https://gnosis-rpc.publicnode.com Tracking:none
URL: https://rpc.tornadoeth.cash/gnosis Tracking:none
```

### What else can be improved in the repository?
- In tests, after removing imports from the build, it looks like tests are on the path to not having a build step. It should be encouraged if possible.
- maybe deprecate `testRpcPerformance()` and introduce a function with a better name. Function names with "test..." prefix are used worldwide in all programming languages to write test functions and not subject source code. 
- do not leak tests into the subject source code. better to use spies or stubs to deal with local storage issue on tests. https://github.com/ubiquity/rpc-handler/blob/30022f441aaeacc2a4c3fcfadf41c3e7784f7f25/src/services/storage-service.ts#L6
